components/ide/code/codehelper/main.go

```diff
@@ -0,0 +1,205 @@
+
+	code := "/ide/bin/gitpod-code"
+	// start code server with --install-extension flag
+	args := []string{code, "--start-server"}
+
+	if os.Getenv("SUPERVISOR_DEBUG_ENABLE") == "true" {
+		args = append(args, "--inspect", "--log=trace")
+	}
+	wsContextUrl := wsInfo.GetWorkspaceContextUrl()
+	if ctxUrl, err := url.Parse(wsContextUrl); err == nil {
+		if ctxUrl.Host == "github.com" {
+			log.Info("ws context url is from github.com, install builtin extension github.vscode-pull-request-github")
+			args = append(args, "--install-builtin-extension", "github.vscode-pull-request-github")
+		}
+	} else {
+		log.WithError(err).WithField("wsContextUrl", wsContextUrl).Error("parse ws context url failed")
+	}
+
+	uniqMap := map[string]struct{}{}
+	extensions, _ := getExtensions(wsInfo.GetCheckoutLocation())
+	for _, ext := range extensions {
+		if _, ok := uniqMap[ext]; ok {
+			continue
+		}
+		uniqMap[ext] = struct{}{}
+		args = append(args, "--install-extension", ext)
+	}
+
+	args = append(args, os.Args...)
+	log.WithField("code", code).WithField("args", args).
+		WithField("cost", time.Now().Local().Sub(startTime).Milliseconds()).
+		Info("run cmd")
+	args = append([]string{code}, args...)
+
+	// use unix.Exec to replace the calling executable in the process tree
+	if err := unix.Exec(code, args, os.Environ()); err != nil {
+		log.WithError(err).Error("unix exec code failed")
+	}
+}
+
+func resolveWorkspaceInfo(ctx context.Context) (*supervisor.ContentStatusResponse, *supervisor.WorkspaceInfoResponse, error) {
+	resolve := func(ctx context.Context, attempt int) (contentStatus *supervisor.ContentStatusResponse, wsInfo *supervisor.WorkspaceInfoResponse, err error) {
+		supervisorAddr := os.Getenv("SUPERVISOR_ADDR")
+		if supervisorAddr == "" {
+			supervisorAddr = "localhost:22999"
+		}
+		supervisorConn, err := grpc.Dial(supervisorAddr, grpc.WithInsecure())
+		if err != nil {
+			err = errors.New("dial supervisor failed: " + err.Error())
+			return
+		}
+		defer supervisorConn.Close()
+		if wsInfo, err = supervisor.NewInfoServiceClient(supervisorConn).WorkspaceInfo(ctx, &supervisor.WorkspaceInfoRequest{}); err != nil {
+			err = errors.New("get workspace info failed: " + err.Error())
+			return
+		}
+		contentStatus, err = supervisor.NewStatusServiceClient(supervisorConn).ContentStatus(ctx, &supervisor.ContentStatusRequest{Wait: true})
+		if err != nil {
+			err = errors.New("get content available failed: " + err.Error())
+		}
+		return
+	}
+	// try resolve workspace info 10 times
+	attempt := 10
+	contentStatus, wsInfo, err := resolve(ctx, attempt)
+	if err != nil {
+		log.WithError(err).Error("resolve workspace info failed")
+		if attempt != 0 {
+			return nil, nil, errors.New("failed with attempt 10 times")
+		}
+		time.Sleep(1 * time.Second)
+		return resolve(ctx, attempt-1)
+	}
+	return contentStatus, wsInfo, err
+}
+
+func getExtensions(repoRoot string) (extensions []string, err error) {
```

Comment:
why do we have error if a client don't use it

maybe we should return to a client and let it log once instead of in each if in this function?